virtual/registry/kv_registry.go

```diff
@@ -221,149 +221,100 @@ func (k *kvRegistry) EnsureActivation(
-		} else {
-			// We need to create a new activation because either:
-			//   1. There is no activation OR
-			//   2. The server the actor is currently activated on has stopped heartbeating OR
-			//   3. The server the actor is currently activated on has blacklisted this actor (most
-			//       likely for balancing reasons)
-			liveServers, err := getLiveServers(ctx, vs, tr)
-			if err != nil {
-				return nil, err
-			}
-			if len(liveServers) == 0 {
-				return nil, fmt.Errorf("0 live servers available for new activation")
-			}
+		// Get references from currently running actor activations
+		refs, activations, err := k.getReferencesFromActivations(ctx, tr, req, isServerIDBlacklisted, vs, ra)
+		if err != nil {
+			return nil, err
+		}
 
-			maxNumHeartbeats := 0
-			for _, s := range liveServers {
-				if s.NumHeartbeats > maxNumHeartbeats {
-					maxNumHeartbeats = s.NumHeartbeats
-				}
-			}
-			if maxNumHeartbeats < k.opts.MinSuccessiveHeartbeatsBeforeAllowActivations {
-				return nil, fmt.Errorf(
-					"maxNumHeartbeats: %d < MinSuccessiveHeartbeatsBeforeAllowActivations(%d)",
-					maxNumHeartbeats, k.opts.MinSuccessiveHeartbeatsBeforeAllowActivations)
-			}
+		ra.Activations = activations
+		isActivated := make(map[string]bool, len(activations))
+		for _, a := range activations {
+			isActivated[a.ServerID] = true
+		}
 
-			// TODO: Update this code once we support configurable replication.
-			var cachedServerID string
-			if len(req.CachedActivationServerIDs) > 0 {
-				cachedServerID = req.CachedActivationServerIDs[0]
-			}
+		// If the desired number of replicas is achieved, return the references
+		if uint64(len(refs)) >= 1+req.ExtraReplicas {
+			return EnsureActivationResult{
+				References:   refs,
+				VersionStamp: vs,
+			}, nil
+		}
 
-			selected := pickServerForActivation(
-				liveServers, k.opts, req.BlacklistedServerID, cachedServerID, !activationExists)
-			serverID = selected.ServerID
-			serverAddress = selected.HeartbeatState.Address
-			serverVersion = selected.ServerVersion
-			currActivation = newActivation(serverID, serverVersion)
+		// We need to create a new activation because the desired number of replicas has not been achieved. This can happen in the following scenarios:
+		//   1. There is no existing activation for the actor.
+		//   2. The server where the actor is currently activated has stopped heartbeating.
+		//   3. The server where the actor is currently activated has blacklisted the actor, typically for load balancing purposes.
+		liveServers, err := getLiveServers(ctx, vs, tr)
+		if err != nil {
+			return nil, err
+		}
+		if len(liveServers) == 0 {
+			return nil, fmt.Errorf("0 live servers available for new activation")
+		}
 
-			ra.Activation = currActivation
-			marshaled, err := json.Marshal(&ra)
-			if err != nil {
-				return nil, fmt.Errorf("error marshaling activation: %w", err)
+		// Find the maximum number of heartbeats among live servers.
+		maxNumHeartbeats := 0
```

Comment:
Could probably make a small helper for lines 273 to 285 just so the code black has a "name" and can be commented